src/Data/Constraint/Deferrable.hs

```diff
@@ -59,12 +64,28 @@ data UnsatisfiedConstraint = UnsatisfiedConstraint String
   deriving (Typeable, Show)
 
 instance Exception UnsatisfiedConstraint
+#if MIN_VERSION_base(4,8,0)
+  where
+    displayException (UnsatisfiedConstraint str) = str
+#endif
 
 -- | Allow an attempt at resolution of a constraint at a later time
 class Deferrable p where
   -- | Resolve a 'Deferrable' constraint with observable failure.
   deferEither :: proxy p -> (p => r) -> Either String r
 
+-- | A good definition of 'deferEither' for a constraint that is never
+-- satisfiable. The error message will indicate what constraint could
+-- not be satisfied.
+deferEitherFailure :: Typeable p => proxy p -> (p => r) -> Either String r
+deferEitherFailure (_ :: proxy p) _ =
+  Left $ "Could not satisfy deferred constraint `" ++ showTypeRep (Proxy :: Proxy p) ++ "’"
+
+-- | A good definition of 'deferEither' for a constraint that is always
+-- satisfiable.
+deferEitherSuccess :: p => proxy p -> (p => r) -> Either String r
```

Comment:
This could be used to implement the `Deferrable ()` instance, right? Why not do so?